packages/contracts/contracts/optimistic-ethereum/ovm/ExecutionManager.sol

```diff
@@ -206,21 +227,72 @@ contract ExecutionManager is ContractResolver {
         bytes memory _callBytes,
         address _fromAddress,
         address _l1MsgSenderAddress,
+        uint _ovmTxGasLimit,
         bool _allowRevert
     )
         public
     {
         StateManager stateManager = resolveStateManager();
 
         require(_timestamp > 0, "Timestamp must be greater than 0");
-        uint _nonce = stateManager.getOvmContractNonce(_fromAddress);
 
         // Initialize our context
-        initializeContext(_timestamp, _queueOrigin, _fromAddress, _l1MsgSenderAddress);
+        initializeContext(_timestamp, _queueOrigin, _fromAddress, _l1MsgSenderAddress, _ovmTxGasLimit);
 
         // Set the active contract to be our EOA address
         switchActiveContract(_fromAddress);
 
+        // Check for individual tx gas limit violation
+        if (_ovmTxGasLimit > gasMeterConfig.OvmTxMaxGas) {
+            // todo handle _allowRevert=true or ideally remove it altogether as it should probably always be false for Fraud Verification purposes.
+            emit EOACallRevert("Transaction gas limit exceeds max OVM tx gas limit");
+            assembly {
+                return(0,0)
+            }
+        }
+
+        // If we are at the start of a new epoch, the current tiime is the new start and curent cumulative gas is the new cumulative gas at stat!
+        if (_timestamp >= gasMeterConfig.GasRateLimitEpochLength + getGasRateLimitEpochStart()) {
+            setGasRateLimitEpochStart(_timestamp);
+            setCumulativeSequencedGasAtEpochStart(
+                getCumulativeSequencedGas()
+            );
+            setCumulativeQueuedGasAtEpochStart(
+                getCumulativeQueuedGas()
+            );
+        }
+
+        // check for gas rate limit
+        // TODO: split up EM somehow?  We are at max stack depth so can't use vars here yet
+        // TODO: make queue origin an enum?  or just configure better?
+        if (_queueOrigin == 0) {
+            if (
+                getCumulativeSequencedGas()
+                - getCumulativeSequencedGasAtEpochStart()
+                + _ovmTxGasLimit
+                >
+                gasMeterConfig.MaxSequencedGasPerEpoch
+            ) {
+                emit EOACallRevert("Transaction gas limit exceeds remaining gas for this epoch and queue origin.");
+                assembly {
+                    return(0,0)
+                }
+            }
+        } else {
+            if (
+                getCumulativeQueuedGas()
+                - getCumulativeQueuedGasAtEpochStart()
+                + _ovmTxGasLimit
+                >
+                gasMeterConfig.MaxQueuedGasPerEpoch
+            ) {
+                emit EOACallRevert("Transaction gas limit exceeds remaining gas for this epoch and queue origin.");
+                assembly {
+                    return(0,0)
+                }
+            }
+        }
+
```

Comment:
Can this get pulled out into a `validateGasLImit(...)` function? Doesn't look like there are any variables that are set from this that are used in `executeTransaction(...)` so it should be cleanly extractable, right?